### PR TITLE
WIP: Marky gh181 surface khr

### PR DESCRIPTION
**Pull request by MarkY-LunarG**
This includes changes needed to allow the ICDs to create and destroy their own KHR_surface objects instead of having the loader generate a generic object they use.  Piers has validated that with his changes, everything appears to work.

### CODE REVIEW

**Comment by MarkY-LunarG** [comment]
BTW, what is the official process for informing IHVs of these kind of changes to the loader/icd interface.  Is this something brought up in the TSG working group, or did you all directly talk to individuals at those companies (Nvidia, AMD, Intel)?

**Comment by courtney-g** [comment]
I would encourage you to contact key stakeholders that are not already involved with the development to make sure they know about it and the potential impact. Other than that we have ChangeLogs for the spec and SDK. The change is backward compatible so it's not absolutely necessary for folks to do anything unless they need that functionality.

File: layers/unique_objects.h [hunk 1]
```diff
@@ -275,13 +275,6 @@ void explicit_DestroyInstance(VkInstance instance, const VkAllocationCallbacks *
 // Handle CreateDevice
 static void createDeviceRegisterExtensions(const VkDeviceCreateInfo *pCreateInfo, VkDevice device) {
     layer_data *my_device_data = get_my_data_ptr(get_dispatch_key(device), layer_data_map);
-    VkLayerDispatchTable *pDisp = get_dispatch_table(unique_objects_device_table_map, device);
-    PFN_vkGetDeviceProcAddr gpa = pDisp->GetDeviceProcAddr;
-    pDisp->CreateSwapchainKHR = (PFN_vkCreateSwapchainKHR)gpa(device, "vkCreateSwapchainKHR");
```
Comment (MarkY-LunarG) [inline layers/unique_objects.h]: Setting up the function pointers is all taken care of in explicit_CreateDevice() -> initDeviceTable() -> layer_init_device_dispatch_table().  It's duplicate code.

File: loader/wsi.c [hunk 2]
```diff
@@ -207,6 +211,32 @@ terminator_DestroySurfaceKHR(VkInstance instance, VkSurfaceKHR surface,
                              const VkAllocationCallbacks *pAllocator) {
     struct loader_instance *ptr_instance = loader_get_instance(instance);
 
+// Android doesn't have to worry about multiple ICD scenario, but the rest do.
+#ifndef VK_USE_PLATFORM_ANDROID_KHR
```
Comment (MarkY-LunarG) [inline loader/wsi.c]: I'm fine with removing all these.  I didn't want to be premature.

File: loader/wsi.c [hunk 3]
```diff
@@ -468,32 +583,83 @@ LOADER_EXPORT VKAPI_ATTR VkResult VKAPI_CALL vkCreateWin32SurfaceKHR(
...
 
-    pIcdSurface->base.platform = VK_ICD_WSI_PLATFORM_WIN32;
-    pIcdSurface->hinstance = pCreateInfo->hinstance;
-    pIcdSurface->hwnd = pCreateInfo->hwnd;
+    pIcdSurface->win_surf.base.platform = VK_ICD_WSI_PLATFORM_WIN32;
+    pIcdSurface->win_surf.hinstance = pCreateInfo->hinstance;
+    pIcdSurface->win_surf.hwnd = pCreateInfo->hwnd;
+
+    // Setup the new sizes and offsets so we can grow the structures in the
+    // future withouth having problems
```
Comment (MarkY-LunarG) [inline loader/wsi.c]: <facepalm>

File: loader/wsi.c [hunk 4]
```diff
@@ -659,32 +874,83 @@ vkCreateWaylandSurfaceKHR(VkInstance instance,
...
     if (pIcdSurface == NULL) {
-        return VK_ERROR_OUT_OF_HOST_MEMORY;
+        vkRes = VK_ERROR_OUT_OF_HOST_MEMORY;
+        goto out;
     }
 
-    pIcdSurface->base.platform = VK_ICD_WSI_PLATFORM_WAYLAND;
-    pIcdSurface->display = pCreateInfo->display;
-    pIcdSurface->surface = pCreateInfo->surface;
+    pIcdSurface->wayland_surf.base.platform = VK_ICD_WSI_PLATFORM_WAYLAND;
```
Comment (MarkY-LunarG) [inline loader/wsi.c]: Seems reasonable.  I'll make a utility function for what I can.

File: loader/wsi.c [hunk 5]
```diff
@@ -468,32 +583,83 @@ LOADER_EXPORT VKAPI_ATTR VkResult VKAPI_CALL vkCreateWin32SurfaceKHR(
...
+                if (VK_SUCCESS != vkRes) {
+                    goto out;
+                }
+            }
+        }
+    }
 
-    *pSurface = (VkSurfaceKHR)pIcdSurface;
+    *pSurface = (VkSurfaceKHR)(pIcdSurface);
+
```
Comment (MarkY-LunarG) [inline loader/wsi.c]: I'm a fan of single entry single exit.  Plus with a lot of the functions, there's so much common clean-up code throughout the loader that's skipped in many cases which has caused a number of leaks, I'm trying to remedy.  I can see your concern, but I think with branch prediction in the CPUs, the logic will be skipped, and this isn't a call that's triggered very often.

File: loader/table_ops.h [hunk 6]
```diff
@@ -540,6 +540,20 @@ loader_lookup_device_dispatch_table(const VkLayerDispatchTable *table,
     if (!strcmp(name, "CmdExecuteCommands"))
         return (void *)table->CmdExecuteCommands;
 
+    if (!strcmp(name, "CreateSwapchainKHR")) {
+        // For CreateSwapChainKHR we need to use the entry and terminator
+        // functions to properly unwrap the SurfaceKHR object.
+        return (void *)vkCreateSwapchainKHR;
```
Comment (courtney-g) [inline loader/table_ops.h]: It seems odd that that specific function doesn't match the pattern of all the others. (Using table->XXX)
Why wouldn't that work here? Is the unwrapping done inside the trampoline code?

File: loader/wsi.c (hunk 2 above)
Comment (courtney-g) [inline loader/wsi.c]: I think it's safe to remove them.

File: loader/wsi.c (hunk 5 above)
Comment (courtney-g) [inline loader/wsi.c]: I'll leave this to your call.

File: loader/table_ops.h (hunk 6 above)
Comment (MarkY-LunarG) [inline loader/table_ops.h]: Because table->XXX always points past the loader for these functions.  It never enters the loader (which goes against what Jon originally told me, but I found through manually debugging).  It's also the only device extension we have to perform any form of unwrapping on.  The others we typically have work to do are instance extensions which already have loader entry/terminator calls.  Remember, I had to do that because the pCreateInfo->surface has to be unwrapped to pass along the ICD version of the surface handle?

File: layers/unique_objects.h (hunk 1 above)
Comment (ianelliottus) [inline layers/unique_objects.h]: Just to clarify, is this a change that's not directly related to the topic of this pull request?

File: loader/loader.c [hunk 7]
```diff
@@ -1634,16 +1640,21 @@ static bool loader_icd_init_entrys(struct loader_icd *icd, VkInstance inst,
...
 #ifdef VK_USE_PLATFORM_XCB_KHR
+    LOOKUP_GIPA(CreateXcbSurfaceKHR, false);
     LOOKUP_GIPA(GetPhysicalDeviceXcbPresentationSupportKHR, false);
 #endif
 #ifdef VK_USE_PLATFORM_XLIB_KHR
+    LOOKUP_GIPA(CreateXlibSurfaceKHR, false);
     LOOKUP_GIPA(GetPhysicalDeviceXlibPresentationSupportKHR, false);
 #endif
 #ifdef VK_USE_PLATFORM_WAYLAND_KHR
+    LOOKUP_GIPA(CreateWaylandSurfaceKHR, false);
```
Comment (ianelliottus) [inline loader/loader.c]: Code is missing here for the MIR entrypoints (both old and new).

File: loader/table_ops.h [hunk 8]
```diff
@@ -259,7 +259,7 @@ static inline void loader_init_device_extension_dispatch_table(
     VkLayerDispatchTable *table = &dev_table->core_dispatch;
     table->AcquireNextImageKHR =
         (PFN_vkAcquireNextImageKHR)gpa(dev, "vkAcquireNextImageKHR");
-    table->CreateSwapchainKHR =
+    table->CreateSwapchainKHR =
```
Comment (ianelliottus) [inline loader/table_ops.h]: Nit: Probably good to remove this trailing-space at the end of a line:-)

File: loader/table_ops.h [hunk 9]
```diff
@@ -540,6 +540,20 @@ loader_lookup_device_dispatch_table(const VkLayerDispatchTable *table,
     if (!strcmp(name, "CmdExecuteCommands"))
         return (void *)table->CmdExecuteCommands;
 
+    if (!strcmp(name, "CreateSwapchainKHR")) {
+        // For CreateSwapChainKHR we need to use the entry and terminator
```
Comment (ianelliottus) [inline loader/table_ops.h]: Nit: Is "entry" the same as "trampoline"?  If so, please change to "trampoline".

File: loader/wsi.c [hunk 10]
```diff
@@ -446,6 +542,40 @@ vkQueuePresentKHR(VkQueue queue, const VkPresentInfoKHR *pPresentInfo) {
...
+        pIcdSurface->platform_size = platform_size;
+        pIcdSurface->non_platform_offset = (uint32_t)(
+            (uint8_t *)(&pIcdSurface->base_size) - (uint8_t *)pIcdSurface);
+        pIcdSurface->entire_size = sizeof(VkIcdSurface);
+
+        if (create_icd_surfs) {
+            pIcdSurface->real_icd_surfaces = loader_instance_heap_alloc(
+                instance, sizeof(VkSurfaceKHR) * instance->total_icd_count,
+                VK_SYSTEM_ALLOCATION_SCOPE_OBJECT);
+            if (pIcdSurface == NULL) {
```
Comment (ianelliottus) [inline loader/wsi.c]: This should be:

if (pIcdSurface->real_icd_surfaces == NULL) {

File: loader/wsi.c [hunk 11]
```diff
@@ -1238,20 +1530,21 @@ VKAPI_ATTR VkResult VKAPI_CALL terminator_CreateDisplayPlaneSurfaceKHR(
         return VK_ERROR_EXTENSION_NOT_PRESENT;
     }
 
-    pIcdSurface = loader_instance_heap_alloc(
-        inst, sizeof(VkIcdSurfaceDisplay), VK_SYSTEM_ALLOCATION_SCOPE_INSTANCE);
+    pIcdSurface =
+        AllocateIcdSurfaceStruct(inst, sizeof(pIcdSurface->display_surf.base),
+                                 sizeof(pIcdSurface->display_surf), false);
```
Comment (ianelliottus) [inline loader/wsi.c]: Does the "false" parameter imply that the VK_KHR_display extension will continue to support the old method into the foreseeable future?  If so, please add an explicit comment.  If not, please tell me more.

File: loader/wsi.c [hunk 12]
```diff
@@ -468,32 +598,66 @@ LOADER_EXPORT VKAPI_ATTR VkResult VKAPI_CALL vkCreateWin32SurfaceKHR(
...
+                    &pIcdSurface->real_icd_surfaces[i]);
+                if (VK_SUCCESS != vkRes) {
+                    goto out;
+                }
+            }
+        }
+    }
 
-    *pSurface = (VkSurfaceKHR)pIcdSurface;
+    *pSurface = (VkSurfaceKHR)(pIcdSurface);
```
Comment (ianelliottus) [inline loader/wsi.c]: Should we initialize pSurface to VK_NULL_HANDLE up above?

File: loader/table_ops.h (hunk 9 above)
Comment (MarkY-LunarG) [inline loader/table_ops.h]: Technically, a trampoline is a function which can be compiled out to optimize a pass-through.  In my terminology, an entry will never go away.

File: loader/wsi.c [hunk 13]
```diff
@@ -956,11 +1254,9 @@ VKAPI_ATTR VkResult VKAPI_CALL terminator_CreateAndroidSurfaceKHR(
     }
 
     // Next, if so, proceed with the implementation of this function:
-    VkIcdSurfaceAndroid *pIcdSurface = NULL;
-
-    pIcdSurface =
+    VkIcdSurfaceAndroid *pIcdSurface =
         loader_instance_heap_alloc(ptr_instance, sizeof(VkIcdSurfaceAndroid),
-                                   VK_SYSTEM_ALLOCATION_SCOPE_INSTANCE);
+                                   VK_SYSTEM_ALLOCATION_SCOPE_OBJECT);
```
Comment (ianelliottus) [inline loader/wsi.c]: Sigh.  Since Android has its own loader, perhaps we should remove the Android-specific code.  Doesn't need to be part of this pull request, but we should do it sometime.

Comment (courtney-g) [inline loader/wsi.c]: Yeah, I had a similar request earlier.

File: loader/table_ops.h (hunk 9 above)
Comment (ianelliottus) [inline loader/table_ops.h]: Sorry, but I'm confused after reading your reply.  Are you using "entry" for what Jon Ashburn, Courtney, et.al. have been calling a "trampoline" throughout the loader code, or is this "entry" function a different type of function?  I'd like to avoid mixing terminology, but am fine with adding a  new term for something different.

File: loader/wsi.h [hunk 14]
```diff
@@ -19,16 +19,49 @@
...
+#endif // VK_USE_PLATFORM_XCB_KHR
+#ifdef VK_USE_PLATFORM_XLIB_KHR
+        VkIcdSurfaceXlib xlib_surf;
+#endif // VK_USE_PLATFORM_XLIB_KHR
+        VkIcdSurfaceDisplay display_surf;
+    };
+    uint32_t base_size; // Size of VkIcdSurfaceBase
+    uint32_t platform_size; // Size of corresponding VkIcdSurfaceXXX
+    uint32_t non_platform_offset; // Start offset to base_size
+    uint32_t entire_size; // Size of entire VkIcdSurface
```
Comment (ianelliottus) [inline loader/wsi.h]: I'm not understanding the purpose of these 4 members.  What are their purpose?

**Comment by ianelliottus** [comment]
I've finished my initial review.  The changes look pretty good!  I see only one serious bug and some minor issues and nits; and some questions.

File: loader/wsi.c (hunk 12 above)
Comment (MarkY-LunarG) [inline loader/wsi.c]: I will, but according to the vulkan spec, in general the values can only be expected to be valid if the return is successful.

File: loader/wsi.c (hunk 13 above)
Comment (MarkY-LunarG) [inline loader/wsi.c]: That's something we can do in the long run.  I've already go a ton of changes in this commit, so I don't want to get too messy.

File: loader/wsi.h (hunk 14 above)
Comment (MarkY-LunarG) [inline loader/wsi.h]: The individual VkIcdSurface structures are all being used by ICDs, currently.  This new code creates the union for these structures, but I added size checks here because I want to allow the structures to change in size without breaking.  This way, if new items are added, we would still know how to skip to get to pertinent data.

File: loader/loader.c (hunk 7 above)
Comment (MarkY-LunarG) [inline loader/loader.c]: Good catch.  I'll fix that.

File: loader/table_ops.h (hunk 9 above)
Comment (MarkY-LunarG) [inline loader/table_ops.h]: My understanding was that a "trampoline" is designed to go away.  But an "entry" is designed to perform initial functionality on entering a library/module/etc.  If I'm wrong, I'm fine with changing the name.  Here's a brief [article](http://www.datchley.name/recursion-tail-calls-and-trampolines/) talking about trampolines.

File: layers/unique_objects.h (hunk 1 above)
Comment (MarkY-LunarG) [inline layers/unique_objects.h]: Good point.  I'll pull it out.

File: loader/wsi.h (hunk 14 above)
Comment (ianelliottus) [inline loader/wsi.h]: Okay, thanks for the clarification!

File: loader/wsi.c (hunk 13 above)
Comment (ianelliottus) [inline loader/wsi.c]: Agreed--best to do this in another pull request.

File: loader/table_ops.h (hunk 9 above)
Comment (ianelliottus) [inline loader/table_ops.h]: I did a little searching, and found different uses for "trampoline" functions (including [).](https://books.google.com/books?id=aJ1av7UFBPwC&pg=PA280&lpg=PA280&dq=thunk+trampoline&source=web&ots=YQcG0nOf7Z&sig=PZA-Pbx19C9Mnw_1myezK4x3QWQ&hl=en&sa=X&oi=book_result&ct=result#v=onepage&q&f=false)  In the end, I'm pretty convinced that you're using "entry" in the way that the loader folks (and Graham Sellers from AMD, and John McDonald from Valve) have used "trampoline", and that you should switch to that in order to avoid confusion:-).

Comment (MarkY-LunarG) [inline loader/table_ops.h]: Will do.
